File: 2024/1.py
import time

from collections import Counter

from utils import readInput

def loadInput(filename: str) -> tuple[list[int], list[int]]:
    lines = readInput(filename)
    ids: list[tuple[int, ...]] = []
    for l in lines:
        ids.append(tuple(map(int, l.split())))
    return tuple(zip(*ids))

def part1(id1: list[int], id2: list[int]) -> None:
    print (f"🎄 Part 1: {sum([abs(x-y)for x, y in zip(sorted(id1), sorted(id2))])}")

def part2(id1: list[int], id2: list[int]) -> None:
    tot = 0
    counts = Counter(id2)
    for n in id1:
        tot += n*counts[n]
    print (f"🎄🎅 Part 2: {tot}")

if __name__ == "__main__":
    title = "Day 1: Historian Hysteria"
    sub = "⛄"*(len(title)//2-1+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs1, inputs2 = loadInput("input_1.txt")
    
    t0 = time.time()
    part1(inputs1, inputs2)
    print ("Time: {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(inputs1, inputs2)
    print ("Time: {:.5f}".format(time.time()-t0))

File: 2024/1.jl
using StatsBase

function loadInput(filename::String)
    lines = readlines(filename)
    ids = [Tuple(parse.(Int64, split(l))) for l in lines]
    return first.(ids), last.(ids) 
end

function part1(id1::Vector{Int64}, id2::Vector{Int64})
    println("🎄 Part 1: $(sum(abs.(x - y) for (x, y) in zip(sort(id1), sort(id2))))")
end

function part2(id1::Vector{Int64}, id2::Vector{Int64})
    tot = 0
    counts = countmap(id2)
    for n in id1
        tot += n * get(counts, n, 0)
    end
    println("🎄🎅 Part 2: $tot")
end

function main()
    title = "Day 1: Historian Hysteria"
    sub = "❄ " ^(length(title) ÷ 2 + 2)

    println()
    println(" $title ")
    println(sub)

    inputs = loadInput("input_1.txt")

    @time part1(inputs...)
    @time part2(inputs...)
end

main()
